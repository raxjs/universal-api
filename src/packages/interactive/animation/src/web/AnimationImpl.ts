import { Animation, AnimationAction, AnimationActionAnimate, AnimationData, AnimationOptions } from '../types';
import { getMergedOptions } from '../common';
import { CONTAINER_NAME } from '@utils/constant';

const ANIMATION_RUNNING = '__UNI_API_ANIMATION_RUNNING__';
const ANIMATION_CANCEL = '__UNI_API_ANIMATION_CANCEL__';
const ANIMATION_ORIGINAL_STYLE = '__UNI_API_ANIMATION_ORIGINAL_STYLE__';

/**
 * normalize unit
 * @param value
 * @param unit
 * @param force
 */
function normalizeUnit(value: number | string, unit: string, force = false): string {
  if (force || typeof value === 'number') {
    return `${value}${unit}`;
  }
  return value;
}

/**
 * handle animation actions queue
 * @param actions
 * @param fn
 * @param onFinish
 */
function handleActionsQueue(
  actions: AnimationAction[],
  fn: (action: AnimationAction, callback: () => void) => any,
  onFinish?: () => any,
) {
  actions = actions.slice();
  const action = actions.shift();
  if (action) {
    fn(action, () => handleActionsQueue(actions, fn, onFinish));
  } else if (onFinish) {
    onFinish();
  }
}


function applyWebAnimation(actions: AnimationAction[], dom?: HTMLElement) {
  // If `dom` is not HTML Node, ignore
  if (!(dom && dom.nodeType === 1)) {
    return;
  }

  // clear the last animation
  if (dom[ANIMATION_RUNNING]) {
    (dom[ANIMATION_CANCEL] || []).forEach((t) => clearTimeout(t));
  } else {
    // cache original style
    dom[ANIMATION_ORIGINAL_STYLE] = {
      transitionProperty: dom.style.transitionProperty,
      transitionDuration: dom.style.transitionDuration,
      transitionDelay: dom.style.transitionDelay,
      transitionTimingFunction: dom.style.transitionTimingFunction,
      transformOrigin: dom.style.transformOrigin,
    };
  }
  dom[ANIMATION_RUNNING] = true;
  dom[ANIMATION_CANCEL] = [];

  /**
   * Apply animation action, return the current cost time
   * @param action
   */
  const applyAction = (action): number => {
    const { transition, transformOrigin } = action.option;

    let transform = '';
    action.animates.forEach((animate) => {
      if (animate.type === 'style') {
        const [property, value] = animate.args;
        dom.style[property] = value;
      } else {
        transform += ` ${animate.type}(${animate.args.join(',')})`;
      }
    });
    if (transform) {
      dom.style.transform = transform;
    }

    dom.style.transitionProperty = 'all';
    dom.style.transitionDuration = `${transition.duration}ms`;
    dom.style.transitionDelay = `${transition.delay}ms`;
    dom.style.transitionTimingFunction = transition.timingFunction;
    dom.style.transformOrigin = transformOrigin;

    const cost = transition.delay + transition.duration;
    return isNaN(cost) ? 0 : cost;
  };

  const saveCancel = (timer: any) => {
    dom[ANIMATION_CANCEL] = dom[ANIMATION_CANCEL] || [];
    dom[ANIMATION_CANCEL].push(timer);
  };

  let delay = 0;
  handleActionsQueue(actions, (action, callback) => {
    saveCancel(
      setTimeout(() => {
        delay = applyAction(action);
        callback();
      }, delay),
    );
  }, () => {
    saveCancel(
      // Restore the style at the end of the animation
      setTimeout(() => {
        Object.assign(dom.style, dom[ANIMATION_ORIGINAL_STYLE] || {});
        dom[ANIMATION_RUNNING] = null;
        dom[ANIMATION_CANCEL] = null;
        dom[ANIMATION_ORIGINAL_STYLE] = null;
      }, delay),
    );
  });
}


export default class AnimationImpl implements Animation {
  private options: AnimationOptions;
  private actions: AnimationAction[];
  private currentTransform: Record<string, AnimationActionAnimate>;
  private currentStepAnimates: AnimationActionAnimate[];

  constructor(options?: AnimationOptions) {
    this.options = getMergedOptions(CONTAINER_NAME.WEB, options);
    this.actions = [];
    this.currentTransform = {};
    this.currentStepAnimates = [];
  }

  export(dom?: HTMLElement): AnimationData {
    const actions = this.actions.slice();
    applyWebAnimation(actions, dom);

    this.actions = [];
    return {
      actions,
    };
  }

  step(options?: AnimationOptions): Animation {
    options = options ? getMergedOptions(CONTAINER_NAME.WEB, options) : this.options;

    this.currentStepAnimates.forEach((animate) => {
      let key = animate.type;
      if (key === 'style') {
        key += `.${animate.args[0]}`;
      }
      this.currentTransform[key] = animate;
    });

    const animates = [];
    Object.keys(this.currentTransform).forEach((key) => {
      animates.push(this.currentTransform[key]);
    });

    this.actions.push({
      animates,
      option: {
        transformOrigin: options.transformOrigin,
        transition: {
          delay: options.delay,
          duration: options.duration,
          timingFunction: options.timingFunction,
        },
      },
    });
    this.currentStepAnimates = [];

    return this;
  }

  opacity(value: number): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['opacity', value],
    });
    return this;
  }

  backgroundColor(value: string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['background-color', value],
    });
    return this;
  }

  width(value: number | string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['width', normalizeUnit(value, 'px')],
    });
    return this;
  }

  height(value: number | string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['height', normalizeUnit(value, 'px')],
    });
    return this;
  }

  left(value: number | string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['left', normalizeUnit(value, 'px')],
    });
    return this;
  }

  right(value: number | string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['right', normalizeUnit(value, 'px')],
    });
    return this;
  }

  top(value: number | string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['top', normalizeUnit(value, 'px')],
    });
    return this;
  }

  bottom(value: number | string): Animation {
    this.currentStepAnimates.push({
      type: 'style',
      args: ['bottom', normalizeUnit(value, 'px')],
    });
    return this;
  }

  rotate(angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'rotate',
      args: [normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  rotate3d(x = 0, y = 0, z = 0, angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'rotate3d',
      args: [x, y, z, normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  rotateX(angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'rotateX',
      args: [normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  rotateY(angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'rotateY',
      args: [normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  rotateZ(angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'rotateZ',
      args: [normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  scale(sx: number, sy: number): Animation {
    if (sx === undefined) sx = 1;
    if (sy === undefined) sy = sx;

    this.currentStepAnimates.push({
      type: 'scale',
      args: [sx, sy],
    });
    return this;
  }

  scale3d(sx = 1, sy = 1, sz = 1): Animation {
    this.currentStepAnimates.push({
      type: 'scale3d',
      args: [sx, sy, sz],
    });
    return this;
  }

  scaleX(scale = 1): Animation {
    this.currentStepAnimates.push({
      type: 'scaleX',
      args: [scale],
    });
    return this;
  }

  scaleY(scale = 1): Animation {
    this.currentStepAnimates.push({
      type: 'scaleY',
      args: [scale],
    });
    return this;
  }

  scaleZ(scale = 1): Animation {
    this.currentStepAnimates.push({
      type: 'scaleZ',
      args: [scale],
    });
    return this;
  }

  translate(tx = 0, ty = 0): Animation {
    this.currentStepAnimates.push({
      type: 'translate',
      args: [normalizeUnit(tx, 'px', true), normalizeUnit(ty, 'px', true)],
    });
    return this;
  }

  translate3d(tx = 0, ty = 0, tz = 0): Animation {
    this.currentStepAnimates.push({
      type: 'translate3d',
      args: [
        normalizeUnit(tx, 'px', true),
        normalizeUnit(ty, 'px', true),
        normalizeUnit(tz, 'px', true),
      ],
    });
    return this;
  }

  translateX(translation = 0): Animation {
    this.currentStepAnimates.push({
      type: 'translateX',
      args: [normalizeUnit(translation, 'px', true)],
    });
    return this;
  }

  translateY(translation = 0): Animation {
    this.currentStepAnimates.push({
      type: 'translateY',
      args: [normalizeUnit(translation, 'px', true)],
    });
    return this;
  }

  translateZ(translation = 0): Animation {
    this.currentStepAnimates.push({
      type: 'translateZ',
      args: [normalizeUnit(translation, 'px', true)],
    });
    return this;
  }

  skew(ax = 0, ay = 0): Animation {
    this.currentStepAnimates.push({
      type: 'skew',
      args: [normalizeUnit(ax, 'deg', true), normalizeUnit(ay, 'deg', true)],
    });
    return this;
  }

  skewX(angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'skewX',
      args: [normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  skewY(angle = 0): Animation {
    this.currentStepAnimates.push({
      type: 'skewY',
      args: [normalizeUnit(angle, 'deg', true)],
    });
    return this;
  }

  matrix(a = 1, b = 0, c = 0, d = 1, tx = 0, ty = 0): Animation {
    this.currentStepAnimates.push({
      type: 'matrix',
      args: [a, b, c, d, tx, ty],
    });
    return this;
  }

  matrix3d(
    a1 = 1, b1 = 0, c1 = 0, d1 = 0,
    a2 = 0, b2 = 1, c2 = 0, d2 = 0,
    a3 = 0, b3 = 0, c3 = 1, d3 = 0,
    a4 = 0, b4 = 0, c4 = 0, d4 = 1,
  ): Animation {
    this.currentStepAnimates.push({
      type: 'matrix3d',
      args: [
        a1, b1, c1, d1,
        a2, b2, c2, d2,
        a3, b3, c3, d3,
        a4, b4, c4, d4,
      ],
    });
    return this;
  }
}

