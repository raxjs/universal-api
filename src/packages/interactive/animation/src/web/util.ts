import { AnimationAction, AnimationOptions } from '../types';
import { normalizeOptions } from '../common';
import { CONTAINER_NAME } from '@utils/constant';

/**
 * normalize unit
 * @param value
 * @param unit
 * @param force
 */
export function normalizeUnit(value: number | string, unit: string, force = false): string {
  if (force || typeof value === 'number') {
    return `${value}${unit}`;
  }
  return value;
}

/**
 * merge default options
 * @param options
 */
export function getDefaultOptions(options?: AnimationOptions): AnimationOptions {
  return normalizeOptions(
    {
      duration: 400,
      timingFunction: 'linear',
      delay: 0,
      transformOrigin: '50% 50% 0',
      ...options,
    },
    CONTAINER_NAME.WEB,
  );
}

/**
 * handle animation actions queue
 * @param actions
 * @param fn
 * @param onFinish
 */
export function handleActionsQueue(
  actions: AnimationAction[],
  fn: (action: AnimationAction, callback: () => void) => any,
  onFinish?: () => any,
) {
  actions = actions.slice();
  const action = actions.shift();
  if (action) {
    fn(action, () => handleActionsQueue(actions, fn, onFinish));
  } else if (onFinish) {
    onFinish();
  }
}
