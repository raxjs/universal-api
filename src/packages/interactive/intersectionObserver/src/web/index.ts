import { CreateIntersectionObserver, OPTIONS, RECT, CallbackOptions } from '../types';

function buildMarginStr(margins: RECT): string {
  return `${Number(margins.top) || 0 }px ${ Number(margins.right) || 0 }px ${ Number(margins.bottom) || 0 }px ${ Number(margins.left) || 0 }px`;
}

function IntersectionObservers(options?: OPTIONS) {
  this.thresholds = (options && options.thresholds) ? options.thresholds : [0];
  this.initialRatio = (options && options.initialRatio) ? options.initialRatio : 0;
  this.selectAll = (options && options.selectAll) ? options.selectAll : false;
  this.callBack = () => {};
  this.observers = [];
}

function relative(root: Element | null, self, margins?: RECT) {
  const options = {
    root,
    rootMargin: margins ? buildMarginStr(margins) : '0px',
    threshold: self.thresholds,
  };
  self.observers.push(new IntersectionObserver((...args) => {
    self.callBack(...args);
  }, options));
}

IntersectionObservers.prototype.relativeTo = function (selector: string, margins?: RECT) {
  relative(document.querySelector(selector), this, margins);
  return this;
};

IntersectionObservers.prototype.relativeToViewport = function (margins?: RECT) {
  relative(null, this, margins);
  return this;
};

IntersectionObservers.prototype.observe = function (targetSelector: string, callback: (res: CallbackOptions) => any) {
  if (this.selectAll) {
    const targets = document.querySelectorAll(targetSelector);
    targets.forEach((target) => this.observers.forEach((observer) => observer.observe(target)));
  } else {
    const target = document.querySelector(targetSelector);
    this.observers.forEach((observer) => observer.observe(target));
  }
  this.callBack = (entries, observer) => {
    entries.forEach((entry) => {
      callback && typeof callback === 'function' && callback({
        boundingClientRect: entry.boundingClientRect,
        intersectionRatio: entry.intersectionRatio,
        intersectionRect: entry.intersectionRect,
        relativeRect: entry.rootBounds,
        time: entry.time,
      });
    });
  };
  return this;
};

IntersectionObservers.prototype.disconnect = function (targetSelector: string, callback: (res: CallbackOptions) => any) {
  this.callBack = () => {};
  this.observers = [];
  this.observers.forEach((observer) => {
    observer.disconnect();
  });
  return this;
};

const createIntersectionObserver: CreateIntersectionObserver = (options?, component?) => {
  return new IntersectionObservers(options);
};

export default createIntersectionObserver;
