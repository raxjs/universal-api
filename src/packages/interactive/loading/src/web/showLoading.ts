import { normalizeShow } from '../common';
import { CONTAINER_NAME } from '@utils/constant';

declare let window: any;
// window.__uni_loadingWin = null;

const clsPrefix = '__universal_loading';

const styles = `.${clsPrefix} {
  background-color: rgba(0, 0, 0, 0.6);
  box-sizing: border-box;
  max-width: 80%;
  color: #ffffff;
  padding: 15px;
  position: fixed;
  left: 50%;
  bottom: 50%;
  border-radius: 4px;
  text-align: center;
  transform: translateX(-50%);
  -webkit-transform: translateX(-50%);
  z-index: 9999;
  outline: none;
}
.${clsPrefix}_circle {
  border: 2px solid rgba(200, 200, 200, 0.5);
  border-radius: 50%;
  position: relative;
  height: 40px;
  width: 40px;
  margin: auto;
}
.${clsPrefix}_circle:after {
  content: "";
  display: block;
  border-radius: 50%;
  width: 100%;
  height: 100%;
  position: absolute;
  top: -2px;
  left: -2px;
  border: 2px solid;
  border-color: transparent;
  border-left-color: #fff;
  box-sizing: content-box;
  animation: ${clsPrefix}_roller 1s infinite linear;
}
.${clsPrefix}_text {
  margin-top: 6px;
  color: #fff;
  line-height: 1.5;
  font-size: 14px;
  text-align: center;
  font-weight: normal;
}
@keyframes ${clsPrefix}_roller {
  0% {
    transform: rotateZ(0);
  }
  100% {
    transform: rotateZ(360deg);
  }
}`.replace(/\n/g, '');

/*
 * @param message {String}
 */
let styleElement: HTMLElement | null = null;

export const showLoading = normalizeShow(({ content = '', success = () => {}, fail = () => {}, complete = () => {} }) => {
  try {
    if (!styleElement) {
      // create a style tag for keyframes
      styleElement = document.createElement('style');
      styleElement.innerHTML = styles;
      document.body.appendChild(styleElement);
    }
    if (!window.__uni_loadingWin) {
      // create loading win
      window.__uni_loadingWin = document.createElement('div');
      window.__uni_loadingWin.className = clsPrefix;
      window.__uni_loadingWin.setAttribute('role', 'alert');
      // support for ARIA, add tabindex for focus
      // https://developer.mozilla.org/zh-CN/docs/Web/HTML/Global_attributes/tabindex
      window.__uni_loadingWin.setAttribute('tabindex', '-1');
      // add a circle element
      const circle = document.createElement('div');
      circle.className = `${clsPrefix}_circle`;
      window.__uni_loadingWin.appendChild(circle);
      // add text element
      const text = document.createElement('div');
      text.className = `${clsPrefix}_text`;
      window.__uni_loadingWin.appendChild(text);

      document.body.appendChild(window.__uni_loadingWin);
    }

    const text = window.__uni_loadingWin.querySelector(`.${clsPrefix}_text`) as HTMLDivElement;
    if (content) {
      text.style.display = 'block';
      text.innerText = content;
    } else {
      text.style.display = 'none';
      text.innerText = '';
    }

    success();
    complete();
  } catch (error) {
    fail();
    complete();
  }
}, CONTAINER_NAME.WEB);

export default showLoading;
