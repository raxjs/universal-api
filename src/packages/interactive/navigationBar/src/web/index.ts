import setNavigationBarTitle from './setNavigationBarTitle';

export {
  setNavigationBarTitle,
};
export default {
  setNavigationBarTitle,
};
