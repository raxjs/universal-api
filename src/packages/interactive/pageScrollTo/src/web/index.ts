import { Options } from '../types';
import { normalize } from '../common';
import { CONTAINER_NAME } from '@utils/constant';

export default normalize((options: Options) => {
  const { selector, duration, success, fail, complete } = options;
  let { scrollTop } = options;
  const rootElement = document.documentElement;
  let top = rootElement.scrollTop;

  if (scrollTop == null) {
    if (!selector) {
      const msg = '`scrollTop` or `selector` is required';
      fail(msg);
      complete(msg);
      return;
    }
    const ele = document.querySelector(selector);
    if (!ele) {
      const msg = `The element cannot be found by the selector \`${selector}\``;
      fail(msg);
      complete(msg);
      return;
    }

    scrollTop = top + ele.getBoundingClientRect().top;
  }

  // Already at the target position
  if (top === scrollTop) {
    success('OK');
    complete('OK');
    return;
  }

  // Scroll immediately
  if (Number(duration) === 0) {
    rootElement.scrollTop = scrollTop;
    success('OK');
    complete('OK');
    return;
  }

  const maxScrollTop = rootElement.scrollHeight - rootElement.clientHeight;
  if (scrollTop < 0) {
    scrollTop = 0;
  } else if (scrollTop > maxScrollTop) {
    scrollTop = maxScrollTop;
  }

  // Scroll distance per 1ms
  const speed = (scrollTop - top) / duration;

  const handleScroll = (startTime: number) => {
    requestAnimationFrame(() => {
      const cost = Date.now() - startTime;
      if (cost < 5) { // throttle
        handleScroll(startTime);
        return;
      }

      top = rootElement.scrollTop;
      let nextTop = top + cost * speed;
      if ((speed > 0 && nextTop > scrollTop) || (speed < 0 && nextTop < scrollTop)) {
        nextTop = scrollTop;
      }

      rootElement.scrollTop = nextTop;
      if (nextTop === scrollTop) {
        success('OK');
        complete('OK');
      } else {
        handleScroll(Date.now());
      }
    });
  };

  // start scrolling
  handleScroll(Date.now());
}, CONTAINER_NAME.WEB);
