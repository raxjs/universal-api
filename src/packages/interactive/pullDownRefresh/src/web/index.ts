import startPullDownRefresh from './startPullDownRefresh';
import stopPullDownRefresh from './stopPullDownRefresh';
import onPullDownRefresh from './onPullDownRefresh';

export {
  onPullDownRefresh,
  startPullDownRefresh,
  stopPullDownRefresh,
};
export default {
  onPullDownRefresh,
  startPullDownRefresh,
  stopPullDownRefresh,
};
