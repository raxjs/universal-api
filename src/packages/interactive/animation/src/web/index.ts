import { Animation, AnimationOptions } from '../types';
import AnimationImpl from './AnimationImpl';

export function createAnimation(options?: AnimationOptions): Animation {
  return new AnimationImpl(options);
}

export default {
  createAnimation,
};
