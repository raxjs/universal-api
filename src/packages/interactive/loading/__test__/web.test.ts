import { sleep, testWebAPI } from '@/utils/__test__/util';
import { DOMUtil } from '@/utils/__test__/web';

testWebAPI('loading', async () => {
  const { showLoading, hideLoading } = require('../src/index');
  await showLoading({
    content: 'abc',
  });
  await sleep(300); // await show
  expect(document.body).toMatchSnapshot();

  await hideLoading();
  await sleep(300); // await hide
  expect(document.body).toMatchSnapshot();
});
