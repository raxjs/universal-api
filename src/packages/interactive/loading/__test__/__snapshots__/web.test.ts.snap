// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Platform API: loading Test container: web 1`] = `
<body>
  <style>
    .__universal_loading {  background-color: rgba(0, 0, 0, 0.6);  box-sizing: border-box;  max-width: 80%;  color: #ffffff;  padding: 15px;  position: fixed;  left: 50%;  bottom: 50%;  border-radius: 4px;  text-align: center;  transform: translateX(-50%);  -webkit-transform: translateX(-50%);  z-index: 9999;  outline: none;}.__universal_loading_circle {  border: 2px solid rgba(200, 200, 200, 0.5);  border-radius: 50%;  position: relative;  height: 40px;  width: 40px;  margin: auto;}.__universal_loading_circle:after {  content: "";  display: block;  border-radius: 50%;  width: 100%;  height: 100%;  position: absolute;  top: -2px;  left: -2px;  border: 2px solid;  border-color: transparent;  border-left-color: #fff;  box-sizing: content-box;  animation: __universal_loading_roller 1s infinite linear;}.__universal_loading_text {  margin-top: 6px;  color: #fff;  line-height: 1.5;  font-size: 14px;  text-align: center;  font-weight: normal;}@keyframes __universal_loading_roller {  0% {    transform: rotateZ(0);  }  100% {    transform: rotateZ(360deg);  }}
  </style>
  <div
    class="__universal_loading"
    role="alert"
    tabindex="-1"
  >
    <div
      class="__universal_loading_circle"
    />
    <div
      class="__universal_loading_text"
      style="display: block;"
    >
      abc
    </div>
  </div>
</body>
`;

exports[`Test Platform API: loading Test container: web 2`] = `
<body>
  <style>
    .__universal_loading {  background-color: rgba(0, 0, 0, 0.6);  box-sizing: border-box;  max-width: 80%;  color: #ffffff;  padding: 15px;  position: fixed;  left: 50%;  bottom: 50%;  border-radius: 4px;  text-align: center;  transform: translateX(-50%);  -webkit-transform: translateX(-50%);  z-index: 9999;  outline: none;}.__universal_loading_circle {  border: 2px solid rgba(200, 200, 200, 0.5);  border-radius: 50%;  position: relative;  height: 40px;  width: 40px;  margin: auto;}.__universal_loading_circle:after {  content: "";  display: block;  border-radius: 50%;  width: 100%;  height: 100%;  position: absolute;  top: -2px;  left: -2px;  border: 2px solid;  border-color: transparent;  border-left-color: #fff;  box-sizing: content-box;  animation: __universal_loading_roller 1s infinite linear;}.__universal_loading_text {  margin-top: 6px;  color: #fff;  line-height: 1.5;  font-size: 14px;  text-align: center;  font-weight: normal;}@keyframes __universal_loading_roller {  0% {    transform: rotateZ(0);  }  100% {    transform: rotateZ(360deg);  }}
  </style>
</body>
`;
